### PR TITLE
ControllerInterface: enable hotplugging on macOS

### PR DESCRIPTION
**Pull request by ligfx**
This change is [Reviewable](https://reviewable.io/reviews/dolphin-emu/dolphin/4362)

### CODE REVIEW

**Comment by endrift** [comment]
@dolphin-emu-bot rebuild

**Comment by lioncash** [comment]
Review status: 0 of 3 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[Source/Core/InputCommon/ControllerInterface/OSX/OSXJoystick.h, line 82 at r1](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4362#-KUMoM5egTPyH3c-eDtJ:-KUMoM5fIanFpddkpVAt:bfw40tp) ([raw file](https://github.com/dolphin-emu/dolphin/blob/43169d897f091e3b1602372241388a2e24eef8f6/Source/Core/InputCommon/ControllerInterface/OSX/OSXJoystick.h#L82)):*
> ```objc
>   std::string GetSource() const override;
> 
> private:
> ```

Please provide proper accessors, rather than just exposing everything like this. Same for the keyboard.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4362)*

**Comment by ligfx** [comment]
Review status: 0 of 5 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[Source/Core/InputCommon/ControllerInterface/OSX/OSXJoystick.h, line 82 at r1](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4362#-KUMoM5egTPyH3c-eDtJ:-KUTxEaWD1ZoeScXJZFV:b-896fix) ([raw file](https://github.com/dolphin-emu/dolphin/blob/43169d897f091e3b1602372241388a2e24eef8f6/Source/Core/InputCommon/ControllerInterface/OSX/OSXJoystick.h#L82)):*
<details><summary><i>Previously, lioncash (Mat M.) wrote…</i></summary>
> Please provide proper accessors, rather than just exposing everything like this. Same for the keyboard.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4362)*

File: Source/Core/InputCommon/ControllerInterface/OSX/OSXJoystick.mm [hunk 1]
```diff
@@ -270,5 +270,10 @@
 {
   return m_name;
 }
+
+bool Joystick::IsSameDevice(const IOHIDDeviceRef other_device) const {
```
Comment (leoetlino) [inline Source/Core/InputCommon/ControllerInterface/OSX/OSXJoystick.mm]: The brace should be on the next line.

File: Source/Core/InputCommon/ControllerInterface/OSX/OSXKeyboard.mm [hunk 2]
```diff
@@ -263,5 +263,10 @@
 {
   return m_name;
 }
+
+bool Keyboard::IsSameDevice(const IOHIDDeviceRef other_device) const {
```
Comment (leoetlino) [inline Source/Core/InputCommon/ControllerInterface/OSX/OSXKeyboard.mm]: Brace on next line

**Comment by endrift** [comment]
@dolphin-emu-bot rebuild

**Comment by ligfx** [comment]
Review status: 0 of 5 files reviewed at latest revision, 3 unresolved discussions.

---

*[Source/Core/InputCommon/ControllerInterface/OSX/OSXJoystick.mm, line 274 at r2](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4362#-KUTxAHyszbH0brJSiVu-r2-274:-KUU8C5_02s8mrw2m4i0:b-896fix) ([raw file](https://github.com/dolphin-emu/dolphin/blob/c1fb0d079884a8ea1f66f5a6fd3279679cffe432/Source/Core/InputCommon/ControllerInterface/OSX/OSXJoystick.mm#L274)):*
<details><summary><i>Previously, leoetlino (Léo Lam) wrote…</i></summary>
> The brace should be on the next line.
</details>
Done.

---

*[Source/Core/InputCommon/ControllerInterface/OSX/OSXKeyboard.mm, line 267 at r2](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4362#-KUTxAHyszbH0brJSiVv-r2-267:-KUU8Dkxi7vMGF9mmMbB:b-896fix) ([raw file](https://github.com/dolphin-emu/dolphin/blob/c1fb0d079884a8ea1f66f5a6fd3279679cffe432/Source/Core/InputCommon/ControllerInterface/OSX/OSXKeyboard.mm#L267)):*
<details><summary><i>Previously, leoetlino (Léo Lam) wrote…</i></summary>
> Brace on next line
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4362)*

**Comment by leoetlino** [comment]
LGTM, but this should be tested.

---

Review status: 0 of 5 files reviewed at latest revision, 2 unresolved discussions.

---

*[Source/Core/InputCommon/ControllerInterface/OSX/OSX.mm, line 145 at r3](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4362#-KUcrI56NEj3-18GFgjS:-KUcrI56NEj3-18GFgjT:b-40za5p) ([raw file](https://github.com/dolphin-emu/dolphin/blob/b4744fc39c4572ed0acc29d4b59058e9ad69a2b0/Source/Core/InputCommon/ControllerInterface/OSX/OSX.mm#L145)):*
> ```XML
> static std::string GetDeviceRefName(IOHIDDeviceRef inIOHIDDeviceRef)
> {
>   NSString* pName = (NSString*)IOHIDDeviceGetProperty(inIOHIDDeviceRef, CFSTR(kIOHIDProductKey));
> ```

This should probably be a reinterpret_cast

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4362#-:-KUcrlKabJNZok-wgwqn:bcsshh3)*

**Comment by ligfx** [comment]
Review status: 0 of 5 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[Source/Core/InputCommon/ControllerInterface/OSX/OSX.mm, line 145 at r3](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4362#-KUcrI56NEj3-18GFgjS:-KUd3__FjAlVpYHXQEeu:b-896fix) ([raw file](https://github.com/dolphin-emu/dolphin/blob/b4744fc39c4572ed0acc29d4b59058e9ad69a2b0/Source/Core/InputCommon/ControllerInterface/OSX/OSX.mm#L145)):*
<details><summary><i>Previously, leoetlino (Léo Lam) wrote…</i></summary>
> This should probably be a reinterpret_cast
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4362)*

**Comment by ligfx** [comment]
> LGTM, but this should be tested.

I've tested on macOS 10.12 by plugging and unplugging two different Xbox One controllers. I haven't seen any issues, and the feature works as expected.

**Comment by leoetlino** [comment]
I meant having another person test this. More testing is always good ;)

---

Reviewed 2 of 5 files at r2, 2 of 2 files at r3, 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[Source/Core/InputCommon/ControllerInterface/OSX/OSX.mm, line 19 at r4](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4362#-KUd77wEKlrllCwtkDFZ:-KUd77wEKlrllCwtkDF_:b-qz3xa5) ([raw file](https://github.com/dolphin-emu/dolphin/blob/e1741da5b7d615cc45280e0f0bf362804091cc35/Source/Core/InputCommon/ControllerInterface/OSX/OSX.mm#L19)):*
> ```XML
> 
> #include <map>
> ```

I missed it in my previous review, but it looks like you may want to `#include <thread>` too (and move these includes to just under the license comment header)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4362#-:-KUd6tGk_1XEmyUPgZGp:b-6pduzr)*

**Comment by ligfx** [comment]
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[Source/Core/InputCommon/ControllerInterface/OSX/OSX.mm, line 19 at r4](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4362#-KUd77wEKlrllCwtkDFZ:-KUjNaRFY3xHH0CMRy2q:b-896fix) ([raw file](https://github.com/dolphin-emu/dolphin/blob/e1741da5b7d615cc45280e0f0bf362804091cc35/Source/Core/InputCommon/ControllerInterface/OSX/OSX.mm#L19)):*
<details><summary><i>Previously, leoetlino (Léo Lam) wrote…</i></summary>
> I missed it in my previous review, but it looks like you may want to `#include <thread>` too (and move these includes to just under the license comment header)
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4362)*

**Comment by ligfx** [comment]
Sure. Can we enumerate what else we're looking to test? Different OS version, different devices, different use cases?

---

Review status: 4 of 5 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4362#-:-KUjNcNmWKxwlorbIK5-:big5ep1)*

**Comment by ligfx** [comment]
Updated to take care of a race condition when `DeInit` is called before the hotplug thread runloop has started.

**Comment by kemenaran** [comment]
Trying to test under Mac OS 10.11, with Xcode 8.0.

I get the following errors:

```
[ 30%] Building CXX object Externals/wxWidgets3/CMakeFiles/wx.dir/src/common/arcall.cpp.o
In file included from /Users/kemenaran/Programmes/dolphin-emu/Source/Core/InputCommon/ControllerInterface/OSX/OSX.mm:19:
/Users/kemenaran/Programmes/dolphin-emu/Source/Core/InputCommon/ControllerInterface/OSX/RunLoopStopper.h:31:43: error: unknown type name 'CFRunLoopMode'; did
      you mean 'CFRunLoopRef'?
  void AddToRunLoop(CFRunLoopRef runloop, CFRunLoopMode mode)
                                          ^~~~~~~~~~~~~
                                          CFRunLoopRef
/System/Library/Frameworks/CoreFoundation.framework/Headers/CFRunLoop.h:19:58: note: 'CFRunLoopRef' declared here
typedef struct CF_BRIDGED_MUTABLE_TYPE(id) __CFRunLoop * CFRunLoopRef;
                                                         ^
In file included from /Users/kemenaran/Programmes/dolphin-emu/Source/Core/InputCommon/ControllerInterface/OSX/OSX.mm:19:
/Users/kemenaran/Programmes/dolphin-emu/Source/Core/InputCommon/ControllerInterface/OSX/RunLoopStopper.h:36:48: error: unknown type name 'CFRunLoopMode'; did
      you mean 'CFRunLoopRef'?
  void RemoveFromRunLoop(CFRunLoopRef runloop, CFRunLoopMode mode)
                                               ^~~~~~~~~~~~~
                                               CFRunLoopRef
/System/Library/Frameworks/CoreFoundation.framework/Headers/CFRunLoop.h:19:58: note: 'CFRunLoopRef' declared here
typedef struct CF_BRIDGED_MUTABLE_TYPE(id) __CFRunLoop * CFRunLoopRef;
                                                         ^
In file included from /Users/kemenaran/Programmes/dolphin-emu/Source/Core/InputCommon/ControllerInterface/OSX/OSX.mm:19:
/Users/kemenaran/Programmes/dolphin-emu/Source/Core/InputCommon/ControllerInterface/OSX/RunLoopStopper.h:34:5: error: no matching function for call to
      'CFRunLoopAddSource'
    CFRunLoopAddSource(runloop, m_source, mode);
    ^~~~~~~~~~~~~~~~~~
/System/Library/Frameworks/CoreFoundation.framework/Headers/CFRunLoop.h:73:16: note: candidate function not viable: cannot convert argument of incomplete type
      'CFRunLoopRef' (aka '__CFRunLoop *') to 'CFStringRef' (aka 'const __CFString *')
CF_EXPORT void CFRunLoopAddSource(CFRunLoopRef rl, CFRunLoopSourceRef source, CFStringRef mode);
               ^
In file included from /Users/kemenaran/Programmes/dolphin-emu/Source/Core/InputCommon/ControllerInterface/OSX/OSX.mm:19:
/Users/kemenaran/Programmes/dolphin-emu/Source/Core/InputCommon/ControllerInterface/OSX/RunLoopStopper.h:39:5: error: no matching function for call to
      'CFRunLoopRemoveSource'
    CFRunLoopRemoveSource(runloop, m_source, mode);
    ^~~~~~~~~~~~~~~~~~~~~
/System/Library/Frameworks/CoreFoundation.framework/Headers/CFRunLoop.h:74:16: note: candidate function not viable: cannot convert argument of incomplete type
      'CFRunLoopRef' (aka '__CFRunLoop *') to 'CFStringRef' (aka 'const __CFString *')
CF_EXPORT void CFRunLoopRemoveSource(CFRunLoopRef rl, CFRunLoopSourceRef source, CFStringRef mode);
               ^
/Users/kemenaran/Programmes/dolphin-emu/Source/Core/InputCommon/ControllerInterface/OSX/OSX.mm:217:51: error: cannot initialize a parameter of type
      'CFRunLoopRef' (aka '__CFRunLoop *') with an lvalue of type 'CFStringRef' (aka 'const __CFString *')
    s_stopper.AddToRunLoop(CFRunLoopGetCurrent(), OurRunLoop);
                                                  ^~~~~~~~~~
/Users/kemenaran/Programmes/dolphin-emu/Source/Core/InputCommon/ControllerInterface/OSX/RunLoopStopper.h:31:57: note: passing argument to parameter 'mode' here
  void AddToRunLoop(CFRunLoopRef runloop, CFRunLoopMode mode)
                                                        ^
/Users/kemenaran/Programmes/dolphin-emu/Source/Core/InputCommon/ControllerInterface/OSX/OSX.mm:219:56: error: cannot initialize a parameter of type
      'CFRunLoopRef' (aka '__CFRunLoop *') with an lvalue of type 'CFStringRef' (aka 'const __CFString *')
    s_stopper.RemoveFromRunLoop(CFRunLoopGetCurrent(), OurRunLoop);
                                                       ^~~~~~~~~~
/Users/kemenaran/Programmes/dolphin-emu/Source/Core/InputCommon/ControllerInterface/OSX/RunLoopStopper.h:36:62: note: passing argument to parameter 'mode' here
  void RemoveFromRunLoop(CFRunLoopRef runloop, CFRunLoopMode mode)
                                                             ^
6 errors generated.
```

**Comment by kemenaran** [comment]
This seems to be because `CFRunLoopMode` is typedef-ed only starting from the SDK for macOS 10.12.

**Comment by ligfx** [comment]
Thanks for the catch @kemenaran, I've updated the PR to use `CFStringRef` instead.

**Comment by leoetlino** [comment]
@dolphin-emu-bot rebuild

**Comment by kemenaran** [comment]
@ligfx thanks! The updated version does compile on my machine now.

I tested:

- Starting a game with a plugged controller : ✅
- Plugging a controller while the game is running : ✅
- Unplugging and plugging again a controller while the game is running : ✅

Thanks for tackling this, the lack of hotplug under macOS has been bothering me for a while!

**Comment by kemenaran** [comment]
NB: this was tested with a Gamecube controller, and an unofficial NGC-to-USB adapter.

**Comment by leoetlino** [comment]
Reviewed 2 of 2 files at r5.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[Source/Core/InputCommon/ControllerInterface/OSX/OSX.mm, line 145 at r5](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4362#-KWYwP4rXXwmMb20hW_j:-KWYwP4rXXwmMb20hW_k:b-3blahk) ([raw file](https://github.com/dolphin-emu/dolphin/blob/7716899e0e461b412fa3e94211cbbb2385701156/Source/Core/InputCommon/ControllerInterface/OSX/OSX.mm#L145)):*
> ```XML
> static std::string GetDeviceRefName(IOHIDDeviceRef inIOHIDDeviceRef)
> {
>   const NSString* pName = reinterpret_cast<const NSString*>(
> ```

Dolphin doesn't use Hungarian notation, so this should be just "name".

---

*[Source/Core/InputCommon/ControllerInterface/OSX/RunLoopStopper.h, line 19 at r5](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4362#-KWYwfySVbU8gGAYW6u2:-KWYwfySVbU8gGAYW6u3:b-8fncu6) ([raw file](https://github.com/dolphin-emu/dolphin/blob/7716899e0e461b412fa3e94211cbbb2385701156/Source/Core/InputCommon/ControllerInterface/OSX/RunLoopStopper.h#L19)):*
> ```objc
> 
> public:
>   RunLoopStopper() : m_runloop{nullptr}
> ```

FYI, you can probably simply have `CFRunLoopRef m_lunloop = nullptr;`, which is more concise.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4362)*

**Comment by ligfx** [comment]
Review status: 2 of 6 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[Source/Core/InputCommon/ControllerInterface/OSX/OSX.mm, line 145 at r5](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4362#-KWYwP4rXXwmMb20hW_j:-KWdqpMIHICsIgtHkHb0:b-896fix) ([raw file](https://github.com/dolphin-emu/dolphin/blob/7716899e0e461b412fa3e94211cbbb2385701156/Source/Core/InputCommon/ControllerInterface/OSX/OSX.mm#L145)):*
<details><summary><i>Previously, leoetlino (Léo Lam) wrote…</i></summary>
> Dolphin doesn't use Hungarian notation, so this should be just "name".
</details>
Done.

---

*[Source/Core/InputCommon/ControllerInterface/OSX/RunLoopStopper.h, line 19 at r5](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4362#-KWYwfySVbU8gGAYW6u2:-KWdqqjMpKLw34pNATqK:b-896fix) ([raw file](https://github.com/dolphin-emu/dolphin/blob/7716899e0e461b412fa3e94211cbbb2385701156/Source/Core/InputCommon/ControllerInterface/OSX/RunLoopStopper.h#L19)):*
<details><summary><i>Previously, leoetlino (Léo Lam) wrote…</i></summary>
> FYI, you can probably simply have `CFRunLoopRef m_lunloop = nullptr;`, which is more concise.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4362)*

**Comment by ligfx** [comment]
Rebased on master.

**Comment by Helios747** [comment]
@dolphin-emu-bot rebuild

**Comment by Helios747** [comment]
Fix my comments and I'll merge it.

---

Review status: 2 of 6 files reviewed at latest revision, 5 unresolved discussions.

---

*[Source/Core/InputCommon/ControllerInterface/OSX/OSX.mm, line 150 at r5](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4362#-KXrcYzXf9LQX8FRdbHp:-KXrcYzXf9LQX8FRdbHq:b3k13yr) ([raw file](https://github.com/dolphin-emu/dolphin/blob/7716899e0e461b412fa3e94211cbbb2385701156/Source/Core/InputCommon/ControllerInterface/OSX/OSX.mm#L150)):*
> ```XML
> }
> 
> static void DeviceRemoval_callback(void* inContext, IOReturn inResult, void* inSender,
> ```

This should be named "DeviceRemovalCallback()" as functions should be in CamelCase as defined in our Contributing.md

---

*[Source/Core/InputCommon/ControllerInterface/OSX/OSX.mm, line 168 at r5](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4362#-KXrc8Qme51wCAJ3nhnU:-KXrc8Qnwt4oBFfUyZhG:besz2t7) ([raw file](https://github.com/dolphin-emu/dolphin/blob/7716899e0e461b412fa3e94211cbbb2385701156/Source/Core/InputCommon/ControllerInterface/OSX/OSX.mm#L168)):*
> ```XML
> }
> 
> static void DeviceMatching_callback(void* inContext, IOReturn inResult, void* inSender,
> ```

This should be named "DeviceMatchingCallback()"

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4362#-:-KXrcsOw_3ShOMw1FPxR:b-fbmbee)*

**Comment by ligfx** [comment]
Review status: 2 of 6 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[Source/Core/InputCommon/ControllerInterface/OSX/OSX.mm, line 150 at r5](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4362#-KXrcYzXf9LQX8FRdbHp:-KXsjNiJzUyMtRNLUzsb:b-896fix) ([raw file](https://github.com/dolphin-emu/dolphin/blob/7716899e0e461b412fa3e94211cbbb2385701156/Source/Core/InputCommon/ControllerInterface/OSX/OSX.mm#L150)):*
<details><summary><i>Previously, Helios747 (Anthony) wrote…</i></summary><blockquote>

This should be named "DeviceRemovalCallback()" as functions should be in CamelCase as defined in our Contributing.md
</blockquote></details>

Done.

---

*[Source/Core/InputCommon/ControllerInterface/OSX/OSX.mm, line 168 at r5](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4362#-KXrc8Qme51wCAJ3nhnU:-KXsjOGB6QN9n4a70Zsh:b-896fix) ([raw file](https://github.com/dolphin-emu/dolphin/blob/7716899e0e461b412fa3e94211cbbb2385701156/Source/Core/InputCommon/ControllerInterface/OSX/OSX.mm#L168)):*
<details><summary><i>Previously, Helios747 (Anthony) wrote…</i></summary><blockquote>

This should be named "DeviceMatchingCallback()"
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4362)*

**Comment by Helios747** [comment]
@dolphin-emu-bot rebuild

**Comment by Helios747** [comment]
Please fix linter before I can merge this.

check contributing.md if you need to know how to make linter happy

**Comment by Helios747** [comment]
@dolphin-emu-bot rebuild

**Comment by ligfx** [comment]
Oof. Of all the things to run afoul of, it's two spaces...

Linter should be fixed.

**Pull request merged**